[ShanHng] iP

ShanHng opened the pull request:
### Babe

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Babe frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST TO USE

It's very simple to set up!
1. Download it 
2. Double-click it 
3. Start adding tasks!

The best part is it is **FREE**

Features:
- [x] Managing Tasks
- [x] Managing Deadlines
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}

```

Commit:
Add Gradle support

ShanHng committed:
Implement Greet, Echo and Exit for Babe

Duke has been renamed as Babe.

Several skeletal functionalties were added:
- Greets user.
- Listens to input from user and repeats after them.
- Terminate the chat session once "bye"is received from user.

ShanHng committed:
Fragment listenAndRepeat() into listen() and print()

To optimize code reuse, listenAndRepeat() has been broken
down into two separate functions:
a) listen()
- Receives user input
b) print()
- Prints user-input word stored in Babe's memory

Also, this change relocates control to exit Babe to the loop in
main() of Babe.java.

ShanHng committed:
Add functionality to store and list multiple items

Babe no longer repeats words entered by user back to them.

It now stores the input into a string array, and list them out in a
numbered list when asked to.

ShanHng committed:
Add "Mark/Unmark as Done" capabilities

Babe allows its users to store a list.

To make a functional to-do list, users should be able to mark the
list items as Done/Undone.

Also, this commit adds checkboxes next the each item on the list.

ShanHng committed:
Allow Babe to detect instructions more efficiently

Prior to this, in order to detect a status change for an item,
Babe will extract the first word of userInput string in a separate
process and check for for "mark" and "unmark".

However, this requires code that caters to each instruction such as
"mark" and "unmark" to ensure that extraction of substring does not
encounter NullPointerException.

By making userInput an ArrayList of String type, and parsing the
string into the ArrayList, to obtain an instruction from a user
input (apart from add item functionality), we simply have to get
the first element from the ArrayList.

Thus, in this commit, we:
- Change userInput from type String to ArrayList
- Remove checkChangeStatus() that detects whether current
instruction asks for mark/unmark
- Add helper function to rebuild user input from the ArrayList

ShanHng committed:
Create Item class for all list items in Babe

In Level-4 (yet to be implemented), the list items in Babe now has a
few different types, namely, ToDos, Deadlines and Events.

Each of these Items keeps track of different data, and has different
String representations.

Thus, in this commit, Item is abstracted from Babe's list as a
standalone abstract class, with each different type of Item as its
subclasses for ease of usage and better encapsulation.

Each Item now stores a String content and additional String data
such as deadlines and start/end date depending on the object. The
Done status is also now recorded in each Item.

ShanHng committed:
Add functionality to add ToDos, Deadlines and Events

Items added to Babe's list can be in either of these forms:
ToDos, Deadlines and Events. Each of them has a different String
representation.

This commit depreciates previous general addItem() function and
introduces functions to add each type of items to the list.

ShanHng committed:
Babe.java: Fix bugs that cause NoSuchElementException during script
test

ShanHng committed:
text-ui-test: Add input.txt and EXPECTED.TXT for semi-automate
testing of Babe

ShanHng committed:
Rename Item class to Task

ShanHng committed:
java/exception: Add NoDescriptionException and
NonsenseInputException

Babe will be more restrictive in the coming updates.

It will not accept a Task with no description, and it only
recognizes a pool of commands.

ShanHng committed:
Babe.java: Add exception handling for NoDescriptionException and
NonsenseInputException

ShanHng committed:
test: Update input.txt and EXPECTED.txt to test error handling

input.txt now includes invalid entries that will trigger
NonsenseInputException and NoDescriptionException.

EXPECTED.TXT now includes messages for the two aforementioned
Exceptions.

ShanHng committed:
chore: Change Babe's memory data type from array to ArrayList

ShanHng committed:
feat: Add delete functionality to Babe

ShanHng committed:
chore: Tidy up code

ShanHng committed:
feat: Add Save feature to Babe

Upon every change in the task list, Babe will write down the changes to
a local file named "babe.txt".

Babe will attempt to load files from babe.txt every time it is created.

If no such file exists, Babe will create a new babe.txt file.

ShanHng committed:
chore: Update .gitignore to ignore babe.txt

ShanHng committed:
feat: Add date formatting to Deadline and Event tasks

The date format is now standardized as DD MMM YYYY. Time will remain as
specified by user.

ShanHng committed:
Revert "chore: Tidy up code"

This reverts commit 0d2e3f2f0c7ef62c4059f0c2ab16690b72c6370a.

ShanHng committed:
Merge branch-Level-7 into master

ShanHng committed:
Merge branch 'master' of https://github.com/ShanHng/cs2103-ip

ShanHng committed:
Merge branch 'master' of https://github.com/ShanHng/cs2103-ip

ShanHng committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Babe.java

heeeyi commented on src/main/java/Babe.java:
```diff
@@ -0,0 +1,448 @@
+import java.io.*;
```
I feel here more appropriate to import the specific packages instead of using whildcard import?

WindyCall replied:
Aggreed! It would be better to explicitly list all classes imported.

heeeyi commented on src/main/java/Babe.java:
```diff
@@ -0,0 +1,448 @@
+    /**
+     * List of Tasks Babe received from the user.
+     */
+    private ArrayList<Task> memory = new ArrayList<>();
```
Here more appropriate to name it as "memories" to show it a collection of items?

heeeyi commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * String icon for marked.
+     */
+    static String MARKED = "[X]";
```
Consider using a common prefix for the static Strings?

ShanHng replied:
That is a good point! Thank you!

heeeyi commented on src/main/java/Babe.java:
```diff
@@ -0,0 +1,448 @@
+                                true);
+                        chatBot.save();
+
+                    } catch (Exception e) {
```
May be a better practice to catch the specific types of Exceptions, instead of catching the general Exception

heeeyi commented on src/main/java/Babe.java:
```diff
@@ -0,0 +1,448 @@
+ * @version 1.0
+ * @since 17 January 2023
+ */
+public class Babe {
```
It is nice that you format the JavaDoc for author, version and time! I like it!

heeeyi reviewed (commented):
In generall, nice and clean codes that strictly follows the coding standard! Keep it up~

ShanHng committed:
refactor: Create Parser, Storage, TaskList, Ui

Closely related code are extracted into each of their own classes
to enhance readability and ease of maintenance.

The following classes are created:
- Ui:       deals with interactions with the user
- Storage:  deals with loading tasks from the file and saving tasks
            in the file
- Parser:   deals with making sense of the user command
- TaskList: contains the task list e.g., it has operations to
            add/delete tasks in the list

ShanHng committed:
refactor: Create babe package to hold main files

The main files include:
Babe.java, Parser.java, TaskList.java, Ui.java and Storage.java

exception package and task package are moved into babe package as
sub-packages

ShanHng committed:
Merge branch 'add-gradle-support'

ShanHng committed:
test: Add junit tests for TaskList and Parser

ShanHng committed:
fix: Fix gradle.build to enable packaging of Babe into a jar file

Babe can be packaged with Shadow plugin in Gradle.
This commit fixes values in build.gradle for shadowJar and application.

ShanHng committed:
feat: Add "find" functionality to Babe

Users can now key in a search word to find Tasks with a description that
contains the search word entirely.

WindyCall commented on src/main/java/babe/Babe.java:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * A TaskList object for this Babe.
+     */
+    private TaskList taskList = new TaskList();
```
Since you've arranged this class inside babe package, would it be better to use protected modifier?

WindyCall commented on src/main/java/babe/Babe.java:
```diff
@@ -0,0 +1,86 @@
+
+                } else {
+                    throw new NonsenseInputException();
+                }
```
For this list of if and else, should using switch/case condition be with higher readability?

WindyCall commented on src/main/java/babe/Babe.java:
```diff
@@ -0,0 +1,86 @@
+                } else {
+                    throw new NonsenseInputException();
+                }
+            } catch (Exception e) {
```
Is it better to catch specific exception instead of this kind of overall catching? There might be some problems if doing so(i.e catching some unexpected errors not handled properly)

WindyCall commented on src/main/java/babe/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     * @param content The description of the ToDo item.
+     */
+    public Task addToDo(String content, boolean toNotify) {
+        ToDo item = new ToDo(content);
```
Is the name item not suitable here? Maybe Task or todoTask make more sense?

WindyCall commented on src/main/java/babe/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     * @param startDate The start date of the Event. May include time too.
+     * @param endDate The end date of the Event. May include time too.
+     */
+    public Task addEvent(String content, String startDate, String endDate, boolean toNotify) {
```
Should the naming for boolean variable start from **is**, **has** etc?

WindyCall reviewed (commented):
Generally very good practice of naming and most of the codes have excellent code quality!!! Keep  it up and nice work!!

ShanHng committed:
doc: Add class headers to most classes within babe package

ShanHng committed:
grunt: Tidy up existing code

Code that does not comply with the underlying coding standard are
tidied.

ShanHng committed:
Merge branch 'branch-Level-9'

ShanHng committed:
Merge branch 'branch-A-JavaDoc'

ShanHng committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/babe/task/Deadline.java
#	src/main/java/babe/task/Event.java
#	src/main/java/babe/task/ToDo.java

ShanHng committed:
feat: add GUI for Babe

ShanHng committed:
feat: Add GUI for Babe

ShanHng committed:
Merge branch 'master' of https://github.com/ShanHng/cs2103-ip

ShanHng committed:
chore: Fix build.gradle for GUI Jar build

ShanHng committed:
Merge branch 'branch-Level-10'

ShanHng committed:
feat: Add assertions for argument of "mark", "unmark" and "delete"

These three commands (i.e. "mark", "unmark" and "delete") receives an
index as their argument, which refers to the index of the task in the
TaskList.

As TaskList is implemented with an ArrayList, to avoid any errors
resulting from index-referencing, it is important to ensure that
the index argument is:
* an integer
* within the range of the ArrayList

Let's
* create a private method in Parser to assert the index to be an
integer
* create a private method in TaskList to assert the index to be within
ArrayList's range

ShanHng committed:
feat: Add functionality to delete multiple tasks at a time

Over time, the task list will accumulate more and more tasks,
especially since Tasks marked as Done will not be automatically removed
at all. Thus, it will be helpful to have a batch delete function.

Having a delete command that can take in more than just one indices
will be useful to the user.

Let's,
* modify the existing delete command such that it can take in 1 or more
indices to facilitate batch deleting of tasks

ShanHng committed:
Merge pull request #2 from ShanHng/branch-A-Assertions

feat: Add assertions for argument of "mark", "unmark" and "delete"

ShanHng committed:
Merge branch 'master' of https://github.com/ShanHng/cs2103-ip

ShanHng committed:
Merge branch 'master' into C-MassOps

# Conflicts:
#	src/main/java/babe/Parser.java
#	src/main/java/babe/TaskList.java

ShanHng committed:
Merge branch 'C-MassOps'

ShanHng committed:
Merge branch 'master' into branch-A-CodeQuality

ShanHng committed:
refactor: Tidy up code to adhere to Code Quality guidelines

ShanHng committed:
Merge pull request #3 from ShanHng/branch-A-CodeQuality

Merge branch 'master' into branch-A-CodeQuality

ShanHng committed:
Merge branch 'master' of https://github.com/ShanHng/cs2103-ip

ShanHng committed:
Update README.md

ShanHng committed:
Update README.md

ShanHng committed:
Merge branch 'master' of https://github.com/ShanHng/cs2103-ip

ShanHng committed:
chore: Add reuse comments as per Reuse policies

ShanHng committed:
chore: Change babe.png to a license-free image

ShanHng committed:
test: Update expected outputs for JUnit tests

ShanHng committed:
bug: Fix bug showing actual exception message instead of intended
response from Babe

ShanHng committed:
fix: Fix bug showing actual exception message instead of intended
response from Babe

ShanHng committed:
Merge branch 'master' of https://github.com/ShanHng/cs2103-ip

ShanHng committed:
fix: Fix a minor typo

ShanHng committed:
doc: Update README.md with Babe's User Guide

A brief User Guide is provided in the README.md.
Also, a product snapshot is uploaded.

ShanHng committed:
fix: Fix text-truncating behavior when displaying long texts

ShanHng committed:
chore: Remove unused imports
